lib/Sema/TypeCheckConcurrency.cpp

```diff
@@ -2696,6 +2688,53 @@ bool ActorIsolationChecker::mayExecuteConcurrentlyWith(
   return false;
 }
 
+bool ActorIsolationChecker::isDistributedThunk(ConcreteDeclRef ref,
+                                               Expr *context,
+                                               bool isInAsyncLetInitializer) {
+  auto *FD = dyn_cast_or_null<AbstractFunctionDecl>(ref.getDecl());
+  if (!(FD && FD->isInstanceMember() && FD->isDistributed()))
+    return false;
+
+  if (!isa<SelfApplyExpr>(context))
+    return false;
+
+  auto actor = getIsolatedActor(cast<SelfApplyExpr>(context)->getBase());
+
+  // If this is a method reference on an isolated or potentially isolated
+  // actor then it cannot be a remote thunk.
+  if (actor.isIsolated() || actor.isPotentiallyIsolated)
+    return false;
+
+  auto *dc = getDeclContext();
+
+  switch (ActorIsolationRestriction::forDeclaration(ref, dc)) {
+  case ActorIsolationRestriction::CrossActorSelf: {
+    // Here we know that the method could be used across actors
+    // and the actor it's used on is non-isolated, which means
+    // that it could be a thunk in async context.
+    return isInAsyncLetInitializer || isInAsynchronousContext();
+  }
+
+  case ActorIsolationRestriction::ActorSelf: {
+    // An instance member of an actor can be referenced from an actor's
+    // designated initializer or deinitializer.
+    if (actor.isActorSelf() && !isInAsyncLetInitializer) {
```

Comment:
I think this check is appropriate for both `ActorSelf` and `CrossActorSelf` because it means that reference is: ```swift
distributed actor {
   init(transport: AnyTranport) {
      _ = self.test()
   }

   distributed func test() -> Int {  42 }
}
```
The code in `TypeCheckConcurrency` isn't clear on that regard...

Reply:
Let's double check with @kavon; this is "just" an actor function call on self from a synchronous init so it'll behave the same as in actors -- which Kavon is working on now

Reply:
These changes look fine to me. The special checking for actor inits are mainly happening in SIL.

Reply:
Thank you!